### ISSUE TITLE
getent failures in 210_dns_cross_hosts_test.sh

### ISSUE THREAD

**Issue by bboreham** [weaveworks/weave#1171]
Getting errors fairly often, but not every time.  For example:

````
$ ./210_dns_cross_hosts_test.sh 
Resolve names across hosts
cff20f82331df0d70d2fb11e92eff15be65e2a5a5ac408c0e952ba8bb761920a
8404284b1069c09ab4b8e2e316c111878e3626cf45a5476cdb49aa3a7f2c4cc8
921e10ad338c21f7602b05bd834801844fdb571d6486e10d92ff39564d2912ce
28d009291e4ada0f5444dd3c2948db135d953f2ed187175f057100ddd753e6bc
827ca410d18cffc7ac0c91167ae0f1d60f77f33604c89ed5405a77e0cc06df30
test #32 "exec_on 192.168.48.11 baz getent hosts bar.weave.local" failed:
	program terminated with code 0 instead of 2
test #42 "exec_on 192.168.48.11 baz getent hosts bar.weave.local" failed:
	program terminated with code 0 instead of 2
2 of 67 tests failed in 102.807s.
$ ./210_dns_cross_hosts_test.sh 
Resolve names across hosts
adc9137baa3f12b20b4368c7ff1d77999b429ff5f885040e4329893a8ddd3627
200a7c2d47acd8a11764a7a0c4b85ee0cdc13b2b9800904122cd26cf4747e9f2
f4d52eae68cb653cf41562a1bee1190fc777b6908ebc2bbcaf6b4cbf82097e18
513b6063b9159d5653c2f3f934d833c479f97076ead8a7def813ca4a5aea7d42
a09583b06c421755065b48e1f7ce37bad80e7fbbcf6b0cc22db5d886d4234770
test #52 "exec_on 192.168.48.11 baz getent hosts bar.weave.local" failed:
	program terminated with code 0 instead of 2
1 of 67 tests failed in 27.148s.
````

**Comment by bboreham** [comment]
Similar in test 260:

````
$ ./260_dns_removal_test.sh
Automatic DNS record removal on container death
576fc4e50dd43944ceebb9d543ee3c1520baf56742fd20c2856daf2a0dc79b4b
069d90844c156838983427a938e4f5c23a9de44dac722b22aac91e5cecc6d775
3bb9c34d2fe6321e58466febb5e8bc40256c4de6ff572845391a646a28a2fab0
10.2.0.43
018acecc5a096aa7e72c15bfa17e340c0d76fb99d252ea79e4360e2bb75c6cf0
c2
test #4 "exec_on 192.168.48.11 c1 getent hosts seetwo.weave.local | grep -q '10.2.0.43'" failed:
	program terminated with code 1 instead of 0
test #5 "exec_on 192.168.48.11 c1 getent hosts 10.2.0.43 | tr -s ' '" failed:
	expected "10.2.0.43 seetwo.weave.local"
	got nothing
2 of 9 tests failed in 84.932s.
````

**Comment by rade** [comment]
Could just be timing. With the mDNS, when registering a name on one host and then querying it on another would be guaranteed to succeed module packet loss. That's not the case with gossipdns.

**Comment by rade** [comment]
The propagation delay should be tiny though, and much shorter than the time it takes for the tests to move between steps. So this is still somewhat surprising.

**Comment by tomwilkie** [comment]
https://circleci.com/gh/weaveworks/weave/1935
```
---= Running 210_dns_cross_hosts_test.sh =---
Resolve names across hosts
60f52e0a7e74327e39367214b35283a7b9312dd586e6a281e1b324f0a95ff109
e4d3e35133549b10e3dbd2cafba7b255c803374bad02556aa0d8aeff4b300c84
a1b6fdf30df60e21c17026de192f15c63189aa93d79685ca4a8420dfd748cec5
e5169b6935b476e0628bb99155680e19112d819a58bcc1bd55183f5df30fb608
43fea0d64f82858463c36ac78c1bddd8e13af13fd07cf01fda74878f7611ae9c
test #12 "exec_on host1-1935-1.us-central1-a.positive-cocoa-90213 baz getent hosts bar.weave.local" failed:
	program terminated with code 0 instead of 2
1 of 67 tests failed in 60.957s.
```

**Comment by tomwilkie** [comment]
The story so far is that a gossip entry gets delayed or not delivered.  When this happens, the connection is closed with "connection reset by peer".  I can't see the connection being closed by the opposite end, and it looks like the periodic gossip is handling this tidying this up nicely.

**Comment by rade** [comment]
I can reproduce this quite reliably using just two hosts.

**Comment by rade** [comment]
I frequently see `getent`s getting stuck for tens of seconds. Curiously it always seems to be the negative ones, e.g. `+ assert_raises 'exec_on 192.168.48.11 baz getent hosts foo.weave.local' 2
`
running `tcpdump -ni docker0 port 53` gives
````
19:46:08.789037 IP 172.17.1.200.40938 > 172.17.42.1.53: 60310+ AAAA? foo.weave.local. (33)
19:46:13.794279 IP 172.17.1.200.40938 > 172.17.42.1.53: 60310+ AAAA? foo.weave.local. (33)
19:46:18.799584 IP 172.17.1.200.43255 > 172.17.42.1.53: 33451+ AAAA? foo.weave.local.weave.local. (45)
19:46:23.802981 IP 172.17.1.200.43255 > 172.17.42.1.53: 33451+ AAAA? foo.weave.local.weave.local. (45)
19:46:28.808328 IP 172.17.1.200.35836 > 172.17.42.1.53: 58784+ A? foo.weave.local. (33)
19:46:33.813539 IP 172.17.1.200.35836 > 172.17.42.1.53: 58784+ A? foo.weave.local. (33)
19:46:38.818878 IP 172.17.1.200.46767 > 172.17.42.1.53: 47529+ A? foo.weave.local.weave.local. (45)
19:46:38.818999 IP 172.17.42.1.37519 > 172.17.1.197.53: 47529+ A? foo.weave.local.weave.local. (45)
19:46:38.821581 IP 172.17.1.197.53 > 172.17.42.1.37519: 47529* 0/0/0 (45)
19:46:38.821627 IP 172.17.42.1.53 > 172.17.1.200.46767: 47529* 0/0/0 (45)
````

That is all from the single `getent`. Notice the big gaps in the timing.

The corresponding weave log is this:

````
DEBU: 2015/07/17 19:46:08.560669 [allocator aa:40:d3:aa:ab:22]: Container 0d049a25c91fe75f4fc22ccc459467efd228332f570648fb56c33b89777de9f1 died; released addresses
DEBU: 2015/07/17 19:46:33.739590 [allocator aa:40:d3:aa:ab:22]: Allocator.OnGossip: 603 bytes
DEBU: 2015/07/17 19:46:38.819373 [nameserver aa:40:d3:aa:ab:22] local request: {MsgHdr:{Id:47529 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:foo.weave.local.weave.local. Qtype:1 Qclass:1}] Answer:[] Ns:[] Extra:[]}
DEBU: 2015/07/17 19:46:38.819622 [nameserver aa:40:d3:aa:ab:22] lookup foo.weave.local.weave.local. -> &[]
DEBU: 2015/07/17 19:46:38.819658 [nameserver aa:40:d3:aa:ab:22] response: {MsgHdr:{Id:47529 Response:true Opcode:0 Authoritative:true Truncated:false RecursionDesired:true RecursionAvailable:true Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:foo.weave.local.weave.local. Qtype:1 Qclass:1}] Answer:[] Ns:[] Extra:[]}
````

i.e. we don't appear to receive any query until the very end.

**Comment by rade** [comment]
It looks like those stalling requests somehow didn't get NAT'ed correctly.

Here's another dump, this time taken with `docker-ns weave tcpdump -w /tmp/dump1.pcap -ni eth0 port 53` and covering some normal requests, then the stalling ones, and then again some normal requests:
````
20:25:33.940327 IP 172.17.42.1.52028 > 172.17.2.88.53: 50166+ AAAA? bar.weave.local. (33)
20:25:33.940708 IP 172.17.2.88.53 > 172.17.42.1.52028: 50166 NXDomain 0/0/0 (33)
20:25:33.941572 IP 172.17.42.1.50230 > 172.17.2.88.53: 41134+ AAAA? bar.weave.local.weave.local. (45)
20:25:33.941839 IP 172.17.2.88.53 > 172.17.42.1.50230: 41134 NXDomain 0/0/0 (45)
20:25:33.942839 IP 172.17.42.1.48263 > 172.17.2.88.53: 38208+ A? bar.weave.local. (33)
20:25:33.943099 IP 172.17.2.88.53 > 172.17.42.1.48263: 38208* 2/0/0 A 10.2.4.2, A 10.2.4.130 (95)
20:25:34.004325 IP 172.17.42.1.50420 > 172.17.2.88.53: 10475+ PTR? 2.4.2.10.in-addr.arpa. (39)
20:25:34.004788 IP 172.17.2.88.53 > 172.17.42.1.50420: 10475* 1/0/0 PTR bar.weave.local. (89)
20:25:34.067281 IP 172.17.42.1.57400 > 172.17.2.88.53: 11215+ PTR? 130.4.2.10.in-addr.arpa. (41)
20:25:34.067986 IP 172.17.2.88.53 > 172.17.42.1.57400: 11215* 1/0/0 PTR bar.weave.local. (93)
20:25:34.476152 IP 172.17.2.91.49377 > 172.17.42.1.53: 9074+ AAAA? foo.weave.local. (33)
20:25:39.481341 IP 172.17.2.91.49377 > 172.17.42.1.53: 9074+ AAAA? foo.weave.local. (33)
20:25:44.486791 IP 172.17.42.1.60746 > 172.17.2.88.53: 10538+ AAAA? foo.weave.local.weave.local. (45)
20:25:44.488074 IP 172.17.2.88.53 > 172.17.42.1.60746: 10538 NXDomain 0/0/0 (45)
````

Notice how the normal requests have the docker bridge IP as the source, whereas the failing ones originate from a container IP and have the docker bridge as the destination IP.

**Comment by rade** [comment]
re "connection reset by peer"...

I cannot see this in a packet capture inside the (supposedly) originating container, but it *does* show up on a capture on that container's host. Here's the latter:

````
02:02:37.532176 IP (tos 0x0, ttl 63, id 7025, offset 0, flags [DF], proto TCP (6), length 311)
    192.168.48.11.6783 > 192.168.48.12.40671: Flags [P.], cksum 0xfc98 (correct), seq 8131:8390, ack 5799, win 813, options [nop,nop,TS val 6151027 ecr 6145856], length 259
02:02:37.532243 IP (tos 0x0, ttl 64, id 64423, offset 0, flags [DF], proto TCP (6), length 40)
    192.168.48.12.40671 > 192.168.48.11.6783: Flags [R], cksum 0xdbd8 (correct), seq 4049487842, win 0, length 0
````
The first packet is one I can see inside the container too, but not the RST that follows. Note a) the time proximity - 67ns, and the higher-by-one ttl of the RST packet. This does suggest that the first packet somehow causes the host to inject the second.

**Comment by rade** [comment]
`conntrack -E` shows the connection getting DESTROYed after ~10 seconds:

````
[1437182967.444465]	    [NEW] tcp      6 120 SYN_SENT src=172.17.1.114 dst=192.168.48.11 sport=37138 dport=6783 [UNREPLIED] src=192.168.48.11 dst=192.168.48.12 sport=6783 dport=37138
[1437182967.444541]	 [UPDATE] tcp      6 60 SYN_RECV src=172.17.1.114 dst=192.168.48.11 sport=37138 dport=6783 src=192.168.48.11 dst=192.168.48.12 sport=6783 dport=37138
[1437182967.444724]	 [UPDATE] tcp      6 432000 ESTABLISHED src=172.17.1.114 dst=192.168.48.11 sport=37138 dport=6783 src=192.168.48.11 dst=192.168.48.12 sport=6783 dport=37138 [ASSURED]
[1437182977.860131]	[DESTROY] tcp      6 src=172.17.1.114 dst=192.168.48.11 sport=37138 dport=6783 src=192.168.48.11 dst=192.168.48.12 sport=6783 dport=37138 [ASSURED]
````

The 10 seconds do not correspond to any obvious conntrack setting.

**Comment by rade** [comment]
> The 10 seconds do not correspond to any obvious conntrack setting.

...and is coincidental; I have seen other times.

The RST packet does not show up on docker0. So it's getting added between docker0 and eth1.

**Comment by awh** [comment]
After much digging:

1. A weave router on `$HOST1` connects to the weave router on `$HOST2` on port `6783`. This process is facilitated by the iptables DNAT and MASQUERADE rules installed by `docker -p`; an outbound conntrack flow for `$HOST2:6783` is established on `$HOST1` in response to the initial `SYN` packet. This flow moves in to the ESTABLISHED state once the three way handshake is complete.
2. Test script starts and kills application containers in a loop. Weave routers communicate name updates via gossip, exercising established conntrack flow.
3. Each time a container is started, Docker creates new veth pair with a [*random*](https://github.com/torvalds/linux/blob/19583ca584d6f574384e17fe7613dfaeadcdc4a6/include/linux/etherdevice.h#L174) MAC address and adds one end to the `docker0` bridge. Crucially, *if this random MAC is numerically lower than the current bridge MAC, the bridge MAC changes automatically to the lower hardware address* (This is standard behaviour of the bridging module; if you have not explicitly set a bridge MAC, it equals the lowest MAC of currently attached slaves)
4. `$HOST1` weave router container has the docker bridge IP as its default gateway. Traffic destined for the router on` $HOST2` is sent via the container's `eth0` interface (which is one end of a veth pair attached to `docker0`) with the destination MAC corresponding to the docker bridge IP as obtained from ARP.
5. At some point the weave router on `$HOST1` attempts to send a packet to `$HOST2` just after the bridge MAC has changed but before ARP in the container netns has caught up. From the perspective of `$HOST1`, the packet is now no longer destined for `docker0`, but whichever application container is on the other end of the ex-lowest MAC veth
6. The kernel on `$HOST1` notices that it is about to forward a packet matching the `$HOST2:6783` conntrack flow to a different interface (e.g. ex-lowest MAC veth on `docker0`) than the one that established the flow (`eth0` on `$HOST1`). [*It deletes the conntrack mapping as a consequence.*](https://github.com/torvalds/linux/blob/19583ca584d6f574384e17fe7613dfaeadcdc4a6/net/ipv4/netfilter/iptable_nat.c#L138)
7. Now the conntrack flow is gone, router traffic from `$HOST1` to `$HOST2` no longer has its source address masqueraded (and is therefore sent from `$HOST1` `eth0` with an unroutable source address when ARP catches up), and router traffic from `$HOST2` to `$HOST1` is met with `RST`
8. Router on `$HOST2` receives `RST` packets and unilaterally tears down the connection to `$HOST1`
9. `$HOST1` can no longer hear from `$HOST2`; eventually times out peer connection, and attempts to reconnect.
10. The connection hickup results in test failure due to the delay in gossip propagation.

Forcing the bridge MAC to a low value fixes the problem:

````
sudo ip link set docker0 address 00:00:00:00:00:01
````

This subtle interplay between bridge MAC semantics, `docker -p` netfilter/iptables rules and connection tracking is almost certainly the cause of #1192, not to mention any number of other weird docker networking heisenbugs. Explanatory blog post and docker bug filing to follow.

**Comment by dpw** [comment]
> (This is standard behaviour of the bridging module; if you have not explicitly set a bridge MAC, it equals the lowest MAC of currently attached slaves)

Well well well.

The relevant code is at https://github.com/torvalds/linux/blob/1ea2d020ba477cb7011a7174e8501a9e04a325d4/net/bridge/br_stp_if.c#L221

Despite the references to STP, I haven't found anything to suggest that STP defines this behaviour.  And the code in that function dates back more or less unchanged to pre-git kernel versions, so it might take a bit of digging to find any rationale.

**Comment by awh** [comment]
Now I wonder if this could be the cause of #1184 too. Larger numbers of transient containers would increase the chance of creating a veth that changes the bridge MAC; the misrouted packet [gets dropped](https://github.com/torvalds/linux/blob/19583ca584d6f574384e17fe7613dfaeadcdc4a6/net/ipv4/netfilter/iptable_nat.c#L140).

**Comment by rade** [comment]
> Now I wonder if this could be the cause of #1184 too.

I was wondering that too. But how would it cause a "no route to host"?

**Comment by awh** [comment]
> But how would it cause a "no route to host"?

By dropping a UDP packet encapsulating an ARP packet.

**Comment by rade** [comment]
but ARP retries. so the next one should go through, since it won't stumble across stale conntrack flows.

**Comment by awh** [comment]
Good point - I've not examined the effect on UDP traffic in detail, but if it's restricted to dropping the occasional packet (rather than getting into some kind of self-reinforcing failure state) then you're right. Wishful thinking on my part perhaps!

**Comment by rade** [comment]
Is there any way to tell if a bridge's MAC was set explicitly vs automatically? Obviously the kernel knows, but is that info visible in any CLI tools?

(I am thinking about what a possible fix will look like; it's easy for us to set the docker0 bridge MAC in the weave script, since we already do this for the weave bridge, but we don't want to change it with every invocation of the script, and, furthermore, if it was preset already then ideally we want to leave it alone altogether)

**Comment by rade** [comment]
> Is there any way to tell if a bridge's MAC was set explicitly vs automatically?

[There is](https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-net):

````
What:		/sys/class/net/<iface>/addr_assign_type
Date:		July 2010
KernelVersion:	3.2
Contact:	netdev@vger.kernel.org
Description:
		Indicates the address assignment type. Possible values are:
		0: permanent address
		1: randomly generated
		2: stolen from another device
		3: set using dev_set_mac_address
````

Now, curiously, on my desktop, which is running  3.16.0-44-generic #59~14.04.1-Ubuntu, and docker 1.7.1, this is set to 3, but on my smoke test VMs, which were running 3.19.0-21-generic #21-Ubuntu and docker 1.7.0, it was set to 1. Upgrading docker to 1.7.1 changed it to 3. I wonder whether https://github.com/docker/libnetwork/pull/315 fixed the issue as a side effect.

**Comment by rade** [comment]
The plot thickens... docker 1.6.2 sets the bridge MAC with ioctl. MAC. So does 1.7.1. But 1.7.0 uses some netlink magic. Which evidently doesn't work. Though I see no evidence that docker were aware of the issue we are seeing; they went back to ioctl for different reasons.

**Comment by rade** [comment]
FYI, our CircleCI test VMs are running docker 1.7.0.

**Comment by rade** [comment]
So I suppose the "fix" here is for us to refuse to run with docker version 1.7.0. We already have version checks in the weave script, but they all check the *client* version; this issue concerns the server versions, so will need some new code.

**Comment by rade** [comment]
> So I suppose the "fix" here is for us to refuse to run with docker version 1.7.0.

Or set the docker0 bridge MAC if we find it to be of type 1 or 2.

**Comment by tomwilkie** [comment]
Let me know if you think I should build another test VM images with 1.7.1

**Comment by rade** [comment]
> Let me know if you think I should build another test VM images with 1.7.1

We've already updated CircleCI to 1.7.1.

**Comment by awh** [comment]
Will be addressed by #1229.

**Comment by rade** [comment]
Closed by #1231.
